### PR TITLE
Fix `tagQuery()` bugs

### LINKED ISSUE #247: tagQuery() class methods need to work with multiple `class` attributes
Ex:

```r
tagQuery(div(class="A", class="B", "text"))$addClass("C")$removeClass("B")
#> `$allTags()`:
#> <div class="A C B">text</div>
#> 
#> `$selectedTags()`: `$allTags()`
```

### LINKED ISSUE #252: `tagQuery()`: `el$attribs` should never be directly accessed
Use `tagGetAttribute()` instead. See `tagQueryClassAdd()`

### LINKED ISSUE #253: `tagQuery()`: Do not rebuild after `$each()` if length of selected items is 0

### LINKED ISSUE #255: `tagQuery()`: Add `$selectedLength()`

### LINKED ISSUE #257: `tagQuery()`: `hasClass()` should only accept single class value

### LINKED ISSUE #258: `tagQuery()`: `hasAttr()` should only accept single attribute name
Similar to `hasClass()`. Then it should return a vector of the same length as the selected items.

### LINKED ISSUE #259: `tagQuery()`: `attrs` methods should be renamed to `attr`

### LINKED ISSUE #260: `tagQuery()`: should remove `emptyAttrs()` method

### PR DESCRIPTION
**Pull request by schloerke**
Fixes #260
Fixes #259
Fixes #258
Fixes #257
Fixes #255
Fixes #253
Fixes #252
Fixes #247

### CODE REVIEW

File: R/tag_query.R [hunk 1]
```diff
@@ -656,37 +662,32 @@ tagQuery_ <- function(
...
-        #' a  class value already exists, remove it.
+        #' * `$toggleClass(class)`: If the a class(es) value is missing, add it. If
+        #' a class(es) value already exists, remove it.
         toggleClass = function(class) {
           tagQueryClassToggle(selected_, class)
           self
         },
         #' * `$hasClass(class)`: Does each selected tag have particular
-        #' class(es)?
+        #' class?
```
Comment (cpsievert) [inline R/tag_query.R:672]: `class` can be a vector, right?

File: R/tag_query.R [hunk 2]
```diff
@@ -1161,6 +1163,20 @@ removeFromSet <- function(set, vals) {
   # removes the call to `unique()` with `setdiff`
   set[match(set, vals, 0L) == 0L]
 }
+isNonConformClassValue <- function(classVal) {
+  is.null(classVal) ||
+  is.na(classVal) ||
```
Comment (cpsievert) [inline R/tag_query.R:1168]:
```suggestion
  anyNA(classVal) ||
```

**Comment by cpsievert** [comment]
After discussion, we're thinking of keeping the vector support for `addClass(class)` and friends as is. Instead, we will just change `$hasAttr(attr)` to `$hasAttrs(attrs)`

File: R/tag_query.R [hunk 3]
```diff
@@ -1143,16 +1148,19 @@ tagQueryClassHas <- function(els, class) {
     return(rep(FALSE, length(els)))
   }
 
-  classes <- getCssClass(class)
+  class <- prepCssClass(class)
+  if (length(class) > 1 || grepl(" ", class, fixed = TRUE)) {
+    stop("Only a single CSS `class` value is allowed")
```
Comment (cpsievert) [inline R/tag_query.R:1153]:
```suggestion
    stop("Only a single CSS `class` value is allowed", call. = FALSE)
```

File: R/tag_query.R [hunk 4]
```diff
@@ -1088,40 +1093,40 @@ tagQueryAttrsRemove <- function(els, attrs) {
...
 }
 # Check if els have attributes
-tagQueryAttrHas <- function(els, attr) {
-  attr <- as_character2(attr)[[1]]
-  if (length(attr) != 1 || !is.character(attr)) {
-    stop("`attr` must be a single character value")
+tagQueryAttrsHas <- function(els, attrs) {
+  attrs <- as_character2(attrs)
+  if ((length(attrs) == 0) || (!is.character(attrs))) {
+    stop("`attrs` must be a character vector")
```
Comment (cpsievert) [inline R/tag_query.R:1103]:
```suggestion
    stop("`attrs` must be a character vector", call. = FALSE)
```

Comment (cpsievert) [inline R/tag_query.R:1103]: Maybe I missed something but I had thought all of the `class` methods we're going to allow a character vector? That seems sensible to me, but if this is the behavior you want I'll have to roll back some of the doc changes I made in

File: R/tag_query.R [hunk 5]
```diff
@@ -1161,6 +1169,20 @@ removeFromSet <- function(set, vals) {
   # removes the call to `unique()` with `setdiff`
   set[match(set, vals, 0L) == 0L]
 }
+isNonConformClassValue <- function(classVal) {
+  is.null(classVal) ||
+  anyNA(classVal) ||
+  is.list(classVal) ||
```
Comment (cpsievert) [inline R/tag_query.R:1175]: Given the other checks I don't think this is necessary

```suggestion
  anyNA(classVal) ||
```

File: R/tag_query.R [hunk 6]
```diff
@@ -1161,6 +1169,20 @@ removeFromSet <- function(set, vals) {
...
+  (!is.character(classVal)) ||
+  length(classVal) == 0
```
Comment (cpsievert) [inline R/tag_query.R:1177]:
```suggestion
  length(classVal) == 0 ||
  (!is.character(classVal)) ||
  anyNA(classVal)
  
```

File: R/tag_query.R (hunk 4 above)
Comment (schloerke) [reply R/tag_query.R:1103]: I'm confused. This is in the `$hasAttrs(attrs)` method.

The class methods still allow vectors.

Comment (cpsievert) [reply R/tag_query.R]: Ahh, my bad, I meant to add that comment here https://github.com/rstudio/htmltools/pull/262/files#diff-3299ef0635ccf99ae3a3ea996dcd596257444b4a034e704d408a921163f42a02R1152

File: R/tag_query.R [hunk 7]
```diff
@@ -630,6 +630,12 @@ tagQuery_ <- function(
           rebuild_()
           newTagQuery(newSelected)
         },
+        #' ### Length
+        #'
+        #' * `$length()`: Number of tags that have been selected.
+        length = function() {
+          tagQueryFindLength(selected_)
```
Comment (cpsievert) [inline R/tag_query.R:637]:
```suggestion
          length(selected_)
```

File: R/tag_query.R [hunk 8]
```diff
@@ -1207,11 +1233,14 @@ tagQueryClassToggle <- function(els, class) {
...
     }
-    el$attribs$class <- joinCssClass(elClasses)
+    tagEnvSetClassAttrib(el, elClasses)
   })
 }
 
 
+tagQueryFindLength <- function(els) {
+  length(els)
+}
```
Comment (cpsievert) [inline R/tag_query.R:1243]:
```suggestion
```

File: R/tag_query.R (hunk 4 above)
Comment (schloerke) [reply R/tag_query.R]: Good catch!

File: R/tag_query.R [hunk 9]
```diff
@@ -1087,40 +1092,40 @@ tagQueryAttrsRemove <- function(els, attrs) {
...
     stop("`class` must resolve to a character value with a length of at least 1")
   }
-  splitCssClass(class)
+  class
+}
+getCssClass <- function(class) {
+  splitCssClass(prepCssClass(class))
 }
 splitCssClass <- function(class) {
   if (length(class) > 1) {
```
Comment (cpsievert) [inline R/tag_query.R:1131]: Should the `strsplit()` below here be `strsplit(class, "\\s+")[[1]]`?

**Review by cpsievert — APPROVED** [review]
LGTM pending last `strsplit()` nitpick

File: R/tag_query.R (hunk 9 above)
Comment (schloerke) [reply R/tag_query.R:1131]: 41b9ee5

**Pull request merged**